numpyro/contrib/nn/auto_reg_nn.py

```diff
@@ -57,3 +60,117 @@ def create_mask(input_dim, hidden_dims, permutation, output_dim_multiplier):
+    (output_shape, params) pair, and `apply_fun` takes params and inputs
+    and applies the layer.
+
+    :param input_dim: the dimensionality of the input
+    :type input_dim: int
+    :param hidden_dims: the dimensionality of the hidden units per layer
+    :type hidden_dims: list[int]
+    :param param_dims: shape the output into parameters of dimension (p_n, input_dim) for p_n in param_dims
+        when p_n > 1 and dimension (input_dim) when p_n == 1. The default is [1, 1], i.e. output two parameters
+        of dimension (input_dim), which is useful for inverse autoregressive flow.
+    :type param_dims: list[int]
+    :param nonlinearity: The nonlinearity to use in the feedforward network such as ReLU. Note that no
+        nonlinearity is applied to the final network output, so the output is an unbounded real number.
+    :type nonlinearity: callable. defaults to ReLU.
+
+    Reference:
+
+    MADE: Masked Autoencoder for Distribution Estimation [arXiv:1502.03509]
+    Mathieu Germain, Karol Gregor, Iain Murray, Hugo Larochelle
+    """
+    def __init__(self, input_dim, hidden_dims, param_dims=[1, 1], nonlinearity=relu):
+        self.input_dim = input_dim
+        self.hidden_dims = hidden_dims
+        self.masked_layers = None
+        self.param_dims = param_dims
+        self.nonlinearity = nonlinearity
+        self.all_ones = (np.array(param_dims) == 1).all()
+
+        # Calculate the indices on the output corresponding to each parameter
+        ends = onp.cumsum(self.param_dims, axis=0)
+        starts = onp.concatenate((np.zeros(1), ends[:-1]))
+        self.param_slices = [slice(int(s), int(e)) for s, e in zip(starts, ends)]
+
+        self.count_params = len(self.param_dims)
+        self.output_multiplier = sum(self.param_dims)
+
+        # Hidden dimension must be not less than the input otherwise it isn't
+        # possible to connect to the outputs correctly
+        for h in self.hidden_dims:
+            if h < self.input_dim:
+                raise ValueError('Hidden dimension must not be less than input dimension.')
+
+    def init_fun(self, rng, input_shape, permutation=None):
+        """
+        :param rng: rng used to initialize parameters
+        :param input_shape: input shape
+        :param permutation: an optional permutation that is applied to the inputs and controls the order of the
+            autoregressive factorization. in particular for the identity permutation the autoregressive structure
+            is such that the Jacobian is upper triangular. By default this is chosen at random.
+        :type permutation: array of ints
+        """
+        if permutation is None:
+            # By default set a random permutation of variables, which is important for performance with multiple steps
+            rng, rng_perm = random.split(rng)
+            self.permutation = onp.array(random.shuffle(rng_perm, onp.arange(self.input_dim)))
```

Comment:
Is there a reason to use `onp` instead of `np`? It also looks like we are converting the DeviceArray to `onp.array` here.

Reply:
if the outer `onp.array` is not there i get errors like 
```
>       var_index[permutation] = onp.arange(input_dim)
E       IndexError: too many indices for array

numpyro/contrib/nn/auto_reg_nn.py:40: IndexError
```
not sure what's going on

Reply:
Hmm..I think the reason might be that `var_index` itself is `onp.array` which doesn't work well with a JAX DeviceArray for some reason. Anyways, I think fixing this might require a few other changes outside this PR, so best to handle this separately.